### ISSUE TITLE
Agent upgrade retry not gets updated with correct agent binary while upgrading and fails after 15 minutes.

### ISSUE THREAD

**Issue by amolnater-qasource** [elastic/elastic-agent#2945]
**Kibana Build details:**
```
VERSION: 8.8.2 BC1
BUILD: 63328
COMMIT: db8521641e9f4c545859a80e19f34066c507a137
Artifact link: https://artifacts.elastic.co/downloads/beats/elastic-agent/elastic-agent-8.8.1-windows-x86_64.zip
```

**Host OS and Browser version:** All, All

**Preconditions:**
1. 8.8.2 BC1 Kibana cloud environment should be available.
2. Lower version 8.8.1 agent should be installed.

**Steps to reproduce:**
1. Navigate to Fleet>Fleet Settings tab.
2. Add invalid agent binary: `https://staging.elastic.co/8.8.2-7d6121e6/downloads/xxxx`.
3. Navigate to Agents tab and trigger upgrade for 8.8.1 version agent.
4. Now immediately update agent binary with the correct url: `https://staging.elastic.co/8.8.2-7d6121e6/downloads/`.
5. However agent doesn't get updated with correct agent binary on retries.
6. Observe agent retries several times, then fails finally after 15 minutes. 

**Expected:**
Agent upgrade retry should get updated with the correct agent binary while upgrading and should upgrade agent to latest version.

**Related:**
https://github.com/elastic/ingest-dev/issues/1995

**Screen Recording:**

https://github.com/elastic/elastic-agent/assets/77374876/fe6ca9cf-5e4a-49dd-a025-5fcb5d527bf8

![1](https://github.com/elastic/elastic-agent/assets/77374876/73611326-68ad-45f0-bbc0-d1d1165eeb23)

![2](https://github.com/elastic/elastic-agent/assets/77374876/b483735d-f451-4a49-8040-d3f9c6ea113f)

**Agent Logs:**
With Debug log level:
[elastic-agent-diagnostics-2023-06-26T07-38-12Z-00.zip](https://github.com/elastic/elastic-agent/files/11865635/elastic-agent-diagnostics-2023-06-26T07-38-12Z-00.zip)

**Note:**
- This issue is reproducible consistently and every time.

**Comment by manishgupta-qasource** [comment]
Secondary review for this ticket is Done

**Comment by amolnater-qasource** [comment]
Hi @jlind23 

Thank you for looking into this.

If that is the case we might need to update the testcase for this created under https://github.com/elastic/ingest-dev/issues/1995#issuecomment-1576439178

**Screenshot:**
![image](https://github.com/elastic/elastic-agent/assets/77374876/1d29251e-9c29-476c-a3f7-266485030841)

Could you please also confirm if the agent can't pick up the latest agent binary once upgrade is triggered, then there is any use case for the agent retry in case of invalid agent binary?

Or we should update the testcase to the blocked internet only.

Please let us know if we are missing anything here.

Thanks!

**Comment by cmacknz** [comment]
Agreed, not a bug but maybe behavior that could be better.

The new URL needs to come in as a new, separate upgrade action. As of 8.8.1 (see https://github.com/elastic/elastic-agent/pull/2776) the agent will retry the URL for the current upgrade action with no opportunity for to change. We would need to add a way to cancel the upgrade and start a new one from the UI.

Did this work in 8.7.1? This might have been some undocumented behavior that worked previously but we accidentally removed.

**Comment by pierrehilbert** [comment]
I don't think so, correct me if I'm wrong @amolnater-qasource but I think you added this test after our discussion for 8.8.1 tests?

**Comment by amolnater-qasource** [comment]
Hi @pierrehilbert @cmacknz 

Yes, we have executed this specific scenario for the first time on 8.8.2 with 8.8.1 agent, where merges for [Agent upgrade retry](https://github.com/elastic/ingest-dev/issues/1995) were available.

However, we have tried adding incorrect URL to the agent binary earlier before 8.8.0, where agent remained stuck in Updating state permanently. For this we raised an issue here https://github.com/elastic/elastic-agent/issues/2508

Please let us know if anything else is required from our end.
Thanks

**Comment by amolnater-qasource** [comment]
Hi Team,

We have revalidated the agent upgrade retry test scenario by blocking the internet and then enabling it back on 8.8.2 BC2 Kibana cloud environment.

**Observations:**
- We triggered the agent upgrade and immediately blocked the internet for 02 minutes.
- We have observed that 8.8.1 agent successfully upgraded to latest version 8.8.2 on re-enabling internet for the Host machine.

**Build details:**
BUILD: 63337
COMMIT: c7c074756dcb999f9fdf4c153c5445a00653fa44

**Agent Logs:**
[elastic-agent-diagnostics-2023-06-27T09-45-29Z-00.zip](https://github.com/elastic/elastic-agent/files/11880316/elastic-agent-diagnostics-2023-06-27T09-45-29Z-00.zip)

Thanks!
